feature(op-node): pre-fetch receipts concurrently

welkin22 opened the pull request:
### Description
I added pre-fetch receipt logic in the https://github.com/bnb-chain/opbnb/pull/57. However, if the L1 endpoint is not functioning well and the interface response time increases, the efficiency of pre-fetch will be significantly reduced. To address this issue, we parallelize the pre-fetch process for multiple future block heights. This allows us to achieve better performance even when the L1 endpoint is not in good condition.


### Rationale

To solve the problem of low L1 endpoint performance, we have added concurrent logic to improve the efficiency of pre-fetching receipts.

### Example

none

### Changes

Notable changes:
* Pre-fetch adds concurrency logic.
* The context passed into GoOrUpdatePreFetchReceipts has been modified to use the background context, as the ctx has a timeout of 3 seconds. We do not currently need a timeout here.

realknown commented on op-node/sources/l1_client.go:
```diff
@@ -151,13 +155,20 @@ func (s *L1Client) GoOrUpdatePreFetchReceipts(ctx context.Context, l1Start uint6
 					}
-					s.log.Debug("pre-fetching receipts", "block", currentL1Block)
+
+					go func(ctx context.Context, blockInfo eth.L1BlockRef) {
```
why not put the line 152 "L1BlockRefByNumber" here as well?

welkin22 replied:
Because when we reach the latest block height, L1BlockRefByNumber also has the effect of making our processing process wait for a period of time. We assume that if we parallelize the process of L1BlockRefByNumber, then when we reach the latest block height, the processing process will not stop, and it will continue to launch new goroutines to try to process blocks that have not yet been generated. 
On the other hand, the performance of the L1BlockRefByNumber interface is not so bad, and it also has its own cache, so there is no need to parallelize the processing.

krish-nr commented on op-node/sources/l1_client.go:
```diff
@@ -151,13 +155,20 @@ func (s *L1Client) GoOrUpdatePreFetchReceipts(ctx context.Context, l1Start uint6
-					if err != nil {
-						s.log.Warn("failed to pre-fetch receipts", "err", err)
-						time.Sleep(200 * time.Millisecond)
+					waitErr := s.preFetchReceiptsRateLimiter.Wait(ctx)
```
Is it a better way to put this ratelimit.wait before  L1BlockRefByNumber call? 
After triggering the rate limiting threshold, can reduce some unnecessary L1BlockRefByNumber calls

welkin22 replied:
The requests for L1BlockRefByNumber should all be necessary:
When we have not reached the latest block height, the request parameters for L1BlockRefByNumber are different every time, and the results obtained are useful.
When we reach the latest block height, continuously requesting L1BlockRefByNumber allows us to process subsequent processes as soon as a new block height appears. If we add a limiter here, we will still need to wait for a period of time before entering the subsequent process when a new block height appears.

krish-nr reviewed (dismissed):
/LGTM

welkin22 commented:
I changed the value of MaxConcurrentRequests from 10 to 20 in https://github.com/bnb-chain/opbnb/pull/100/commits/ab8dcdf9fe85d5ecffa5b28954aef39a33274c34 and modified the limiter in GoOrUpdatePreFetchReceipts to be half of MaxConcurrentRequests. This is because if GoOrUpdatePreFetchReceipts takes up all of MaxConcurrentRequests, other places that need to request will be throttled. See the code here: https://github.com/bnb-chain/opbnb/blob/e62988aacf1c068650e1ae80001d1fb7da9f141a/op-node/sources/limit.go#L19

@krish-nr PTAL

aweneagle commented on op-node/sources/l1_client.go:
```diff
@@ -151,13 +155,20 @@ func (s *L1Client) GoOrUpdatePreFetchReceipts(ctx context.Context, l1Start uint6
 						time.Sleep(3 * time.Second)
```
It seams we can improve a little bit of performance here. The 3 seconds sleep blocks the handling of `s.preFetchReceiptsStartBlockChan` as well, so we can combine these two channel events into one : `s.preFetchReceiptsStartBlockChan` and `rateLimiter`, with a individual go routine to handle  the `rateLimiter` and its 3 seconds sleep.
